File: lint-arwaky/archive/taxonomy/plugin_manager_error.py
"""plugin_manager_error — Plugin management domain error types."""

from pydantic import BaseModel, ConfigDict

from .error_code_vo import ErrorCode
from .error_value_vo import (
    ErrorMessage,
    Cause,
)
from .adapter_name_vo import AdapterName


class PluginError(BaseModel):
    """General failure during plugin discovery or registration."""

    model_config = ConfigDict(frozen=True)

    message: ErrorMessage
    error_code: ErrorCode | None = None
    cause: Cause | None = None

    def __str__(self):
        code = f" [{self.error_code}]" if self.error_code else ""
        return f"Plugin Error{code}: {self.message}"


class DiscoveryError(PluginError):
    """Failed to discover plugins via entry points."""
    pass


class RegistrationError(PluginError):
    """Failed to register a specific adapter."""
    adapter_name: AdapterName | None = None

    def __str__(self):
        target = f" for '{self.adapter_name}'" if self.adapter_name else ""
        return f"Registration Error{target}: {self.message}"

File: lint-arwaky/src-rust/taxonomy/plugin_manager_error.rs
use serde::{Serialize, Deserialize};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PluginError {
    pub message: ErrorMessage,
    #[serde(default)]
    pub error_code: Option<ErrorCode>,
    #[serde(default)]
    pub cause: Option<Cause>,
}

impl PluginError {
    pub fn new(message: ErrorMessage) -> Self {
        Self { message, error_code: None, cause: None }
    }
}

impl std::fmt::Display for PluginError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let code = self.error_code.as_ref().map(|c| format!(" [{}]", c)).unwrap_or_default();
        write!(f, "Plugin Error{}: {}", code, self.message)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DiscoveryError {
    #[serde(flatten)]
    pub base: PluginError,
}

impl DiscoveryError {
    pub fn new(message: ErrorMessage) -> Self {
        Self { base: PluginError::new(message) }
    }
}

impl std::fmt::Display for DiscoveryError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.base)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RegistrationError {
    #[serde(flatten)]
    pub base: PluginError,
    #[serde(default)]
    pub adapter_name: Option<AdapterName>,
}

impl RegistrationError {
    pub fn new(message: ErrorMessage) -> Self {
        Self { base: PluginError::new(message), adapter_name: None }
    }
}

impl std::fmt::Display for RegistrationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let target = self.adapter_name.as_ref().map(|a| format!(" for '{}'", a)).unwrap_or_default();
        write!(f, "Registration Error{}: {}", target, self.base.message)
    }
}
